### ISSUE TITLE
Bug in keras.layers.DepthwiseConv2D when using stride and dilation

### ISSUE THREAD

**Issue by demmerichs** [tensorflow/tensorflow#31514]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes (attached)
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.13.1
- Python version: 3.6.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.0/7.4.2
- GPU model and memory: Nvidia Quadro P2000 with 4GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
When executing the following lines of code, I get an assertion, because the result is not as expected, but actually the values [-29, -19, 1]. I would see how the -19 would be correct if I had a stride of 1 along the y-axis, however, then it would not match up again with the third number being 1 instead of -4. So it is not only that either dilation or stride is ignored, but actually providing for both non-trivial values results in some strange behavior, which I think is actually a bug.
```
import numpy as np
import tensorflow as tf

kernel = np.array([5, 5], dtype=np.float32).reshape([1, 2, 1, 1])
bias = np.array([-4], dtype=np.float32)
img = np.array([-2, -1, -3, -2, 3, -1, -2, -1, 2], dtype=np.float32).reshape([1, 1, 9, 1])

strides = (2, 2)
dilation = (1, 2)

depthconv2d_layer = tf.keras.layers.DepthwiseConv2D(
    depth_multiplier=1,
    kernel_size=(1, 2),
    strides=strides,
    dilation_rate=dilation,
    padding='valid',
    depthwise_initializer=lambda *args, **kwargs: tf.constant(kernel),
    bias_initializer=lambda *args, **kwargs: tf.constant(bias),
)
depthconv2d_output = depthconv2d_layer(tf.constant(img))

with tf.Session() as s:
    s.run(tf.global_variables_initializer())
    result = s.run(depthconv2d_output)
    print(result)
    assert np.all(result == np.array([-29, -4, 1, -4]).reshape(1, 1, 4, 1))
```
**Describe the expected behavior**
No assertion, as I computed the values by hand and expect them to be right.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Other info / logs**
Nothing else.

**Comment by WindQAQ** [comment]
Seems that this is not a valid argument combination, but TF does not try to raise the assertion or print some warning message.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/convolutional.py#L1700-L1705
>> strides: An integer or tuple/list of 2 integers,
      specifying the strides of the convolution along the height and width.
      Can be a single integer to specify the same value for
      all spatial dimensions.
      Specifying any stride value != 1 is incompatible with specifying
      any `dilation_rate` value != 1.

**Comment by DavidS3141** [comment]
I was aware of this documentation, however this is only noted on the regular convolutions. For the depthwise convolution this documentation (and the raising of an exception) is either missing, or the case is not correctly implemented! I assumed the second case is true, because depthwise convolutions are probably handled differently concerning GPU optimization which might not yield the restriction of being able to specify only one of stride or dilation. Also I find it too coincidental that the docs and the exception is missing.

**Comment by jvishnuvardhan** [comment]
I was able to reproduce the issue with 1.13.1 and even with `tf-nightly`. [Here](https://colab.sandbox.google.com/gist/jvishnuvardhan/4057ece04f24fa13db3c57c3bb36bad0/tf_31514.ipynb) is the gist. Thanks!

**Comment by tanzhenyu** [comment]
@DavidS3141 @WindQAQ   Sorry for the delay! The link is outdated: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/convolutional.py#L1700-L1705

Do you have a new link to what it means by "Seems that this is not a valid argument combination, but TF does not try to raise the assertion or print some warning message"?

**Comment by DavidS3141** [comment]
It is said here:
https://www.tensorflow.org/api_docs/python/tf/keras/layers/Conv2D
under the argument description for dilation rate!

**Comment by tanzhenyu** [comment]
> It is said here:
> https://www.tensorflow.org/api_docs/python/tf/keras/layers/Conv2D
> under the argument description for dilation rate!

The `dilation_rate` is not an argument for DepthwiseConv2D, right?

**Comment by DavidS3141** [comment]
Yes that is true, quite a lot changed of course since TF 1.13.1 which I used in August when I opened this issue. Then the argument was still documented for that version and I opened this issue. TF 1.15 API docs has also no mention of this argument. Not sure what happens with these newer versions when passing the `dilation_rate` as argument to the Depthwise Conv2D

**Comment by tanzhenyu** [comment]
> Yes that is true, quite a lot changed of course since TF 1.13.1 which I used in August when I opened this issue. Then the argument was still documented for that version and I opened this issue. TF 1.15 API docs has also no mention of this argument. Not sure what happens with these newer versions when passing the `dilation_rate` as argument to the Depthwise Conv2D

Yeah it was never part of the API, but it works as intended. I think we should add it to _init_.
On the other hand, I tried to do this through:
```python
tf.nn.depthwise_conv2d(tf.constant(img), tf.reshape(tf.constant([5., 5.]), (1, 2, 1, 1)), strides=(1, 2, 2, 1), padding='VALID', data_format='NHWC', dilations=(1, 2))
```
It looks like the result is the same as the keras layer. I think this is probably a low level bug instead of a tf.keras bug then?

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
